scrapy_splash/response.py

```diff
@@ -173,15 +173,19 @@ def _load_from_json(self):
         if 'url' in self.data:
             self._url = self.data['url']
 
+        # response.headers
+        if 'headers' in self.data:
+            self.headers = headers_to_scrapy(self.data['headers'])
+
         # response.body
         if 'body' in self.data:
             self._body = base64.b64decode(self.data['body'])
-            self._cached_ubody = self._body.decode(self.encoding)
+            self._cached_ubody = self._body.decode(TextResponse(url=self.url,
+                                                                headers=self.headers,
+                                                                body=self._body).encoding)
+            content_type = self.headers.get(b"Content-Type", b"text/plain; charset=utf-8").split(b";")[0]
```

Comment:
I think charset is not the only thing that can follow the ``;`` separator in content-type, one example is ``Content-Type: multipart/form-data; boundary=something``, so I think it would be better to replace only the ``charset=X`` part.

Reply:
A more general question - as I understand, the problem is that when splash renders a non-utf8 page, the body is encoded using original encoding, right? Then another option to solve this problem would be to preserve both the headers and body as is, like done without splash - is this option worse than decoding body and manipulating headers?

Reply:
Rendering of non-utf8 pages works fine. Browser handles encoding. We always get utf-encoded text. That's why we need to strictly set `charset=utf-8` in case of html.
Problem is when we do not render non-utf8 page at browser and use `splash:http_get`.

>would be to preserve both the headers and body as is
I don't understand how this can be achieved not changing current(expected by users) behaviour?

To add special way to handle this type of requests in spider you need to copy-paste all corresponding code or monkey patch existent one. From outher point of view, this changes does not affect current behaviour and add strictness to way encoding handled in middleware.